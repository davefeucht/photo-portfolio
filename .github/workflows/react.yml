# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: React CI/CD

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Install Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}

    - name: Enable corepack
      run: corepack enable

    - name: Install dependencies
      run: yarn install

    - name: Build project
      run: npm run build --if-present

    - name: Run tests
      run: npm test
