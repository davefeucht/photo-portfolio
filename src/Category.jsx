/*****************
* Category component displays the posts for a particular category.
* Props:
*  id: PropTypes.number,
*  name: PropTypes.string,
*  site: PropTypes.string,
*  showAllPosts: PropTypes.bool,
*  clickCategory: PropTypes.func
*****************/

import React from "react";
import axios from "axios";
import Posts from "./Posts.jsx";
import Post from "./Post.jsx";
import PropTypes from "prop-types";

class Category extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      categoryPost: {},
      errorMsg: ""
    };
  }

  _getCategoryPost(categoryId) {
    let getCategoryPostURI = "http://" + this.props.site + "/wp-json/wp/v2/posts?categories=" + categoryId;
    axios.get(getCategoryPostURI)
      .then(res => {
        if(res.data[0] !== undefined) {
          const categoryPost = res.data[0];
          this.setState({categoryPost});
        }
      }, error => {
        const errorMsg = "Did not work: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  }

  _openCategory() {
    this.setState({showAllPosts: !this.state.showAllPosts}); 
    this.props.clickCategory(this.props.id, this.props.name);
  }

  componentWillMount() {
    this._getCategoryPost(this.props.id);
  }

  render() {
    let postBody = "";
    let categoryTitle = <div className="category-title-clickable">{this.props.name}</div>;
    let categorySubtitle = "";
    if(!this.props.showAllPosts) {
      if(!(Object.keys(this.state.categoryPost).length === 0 && this.state.categoryPost.constructor === Object)) {
        postBody = <Post key={this.state.categoryPost.id.toString()} category={this.props.id} categoryName={this.props.name} title={this.state.categoryPost.title.rendered} id={this.state.categoryPost.id} image={this.state.categoryPost.featured_media} context="category-image" site={this.props.site} />;
      }
    }
    else {
      categoryTitle = <div className="category-title">{this.props.name}</div>;
      categorySubtitle = <div className="category-subtitle">Back to Categories</div>; 
      postBody = <Posts site={this.props.site} category={this.props.id} />;
    }
    return ( 
      <div className="category" onClick={this._openCategory.bind(this)}>
        <div>
          {categoryTitle}
          {categorySubtitle}
        </div>
        {postBody}
      </div>
    );
  }

}

Category.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  site: PropTypes.string,
  showAllPosts: PropTypes.bool,
  clickCategory: PropTypes.func
};

export default Category;
