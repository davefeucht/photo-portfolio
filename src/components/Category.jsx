/*****************
* Category component displays the posts for a particular category.
* Props:
*  id: PropTypes.number,
*  name: PropTypes.string,
*  site: PropTypes.string,
*  clickCategory: PropTypes.func
*****************/

import React from "react";
import axios from "axios";
import Posts from "./Posts.jsx";
import CategoryHeader from "./CategoryHeader.jsx";
import StyledCategory from "./styledComponents/StyledCategory.jsx";
import PropTypes from "prop-types";

export default class Category extends React.Component {

  static propTypes = {
    id: PropTypes.number,
    name: PropTypes.string,
    site: PropTypes.string,
    clickCategory: PropTypes.func
  };

  state = {
    categoryPost: {},
    errorMsg: ""
  };

  //Function to get the 'main' post for the category
  _getCategoryPost(categoryId) {
    let getCategoryPostURI = "http://" + this.props.site + "/wp-json/wp/v2/posts?categories=" + categoryId;
    axios.get(getCategoryPostURI)
      .then(res => {
        if(res.data[0] !== undefined) {
          const categoryPost = res.data[0];
          this.setState({categoryPost});
        }
      })
      .catch(error => {
        const errorMsg = "Could not get posts for this category: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  }

  //When the component is about to mount, get the main post for the category
  componentWillMount() {
    this._getCategoryPost(this.props.id);
  }

  render() {
    return ( 
      <StyledCategory>
        <CategoryHeader id={this.props.id} name={this.props.name} clickCategory={this.props.clickCategory.bind(this)} />
        <Posts site={this.props.site} category={this.props.id} />
      </StyledCategory>
    );
  }
}