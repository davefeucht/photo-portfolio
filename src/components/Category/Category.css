.category {
  min-height: 100%;
  display: grid;
  grid-template-columns: 100%;
  grid-template-areas: "category-header" "category-content";
  grid-template-rows: .25fr 4fr;
}