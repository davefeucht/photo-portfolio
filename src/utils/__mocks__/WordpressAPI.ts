const imageUrl = 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg';

const siteInfo = {
    siteName: 'Through a Pinhole',
    siteUrl: 'throughapinhole.com'
};

const categories = [
    {
        id: 23,
        count: 6,
        description: 'Category of other things',
        link: '',
        name: 'Some other category',
        slug: 'smocg',
        taxonomy: '',
        parent: 12,
        meta: {}
    },
    {
        id: 35,
        count: 3,
        description: 'Category of things',
        link: '',
        name: 'Some category',
        slug: 'smcg',
        taxonomy: '',
        parent: 23,
        meta: {}
    }
];

const posts = [
    {
        id: 887,
        link: '',
        modified: '',
        slug: '',
        type: '',
        featured_media: 142,
        categories: [220],
        title: {
            rendered: 'Some post'
        },
        content: {
            rendered: 'Some content',
            protected: false
        },
        thumbnail_image: 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg'
    },
    {
        id: 889,
        featured_media: 145,
        categories: [220],
        title: {
            rendered: 'Some other post'
        },
        thumbnail_image: 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg'
    }
];

const pages = [
    {
        id: 150,
        content: {
            rendered: 'Contact me!'
        },
        title: {
            rendered: 'Contact'
        }
    }
];

const tagNames = ['something', 'somethingelse'];

export default class WordpressAPI {
    siteUrl: string;

    constructor(siteUrl: string) {
        this.siteUrl = siteUrl;
    }

    async getPostThumbnail() {
        return Promise.resolve(imageUrl);
    }

    async getSiteInfo() {
        return Promise.resolve(siteInfo.siteName);
    }

    async getCategories() {
        return Promise.resolve(categories);
    }

    async getCategoryImage() {
        return Promise.resolve(imageUrl);
    }

    async getPosts() {
        return Promise.resolve(posts);
    }

    async getPost() {
        return Promise.resolve(posts[0]);
    }

    async getCategoryInfo() {
        return Promise.resolve(categories[0]);
    }

    async getPostImage() {
        return Promise.resolve(imageUrl);
    }

    async getTagNames() {
        return Promise.resolve(tagNames);
    }

    async getPages() {
        return Promise.resolve(pages);
    }

    async getPage() {
        return Promise.resolve(pages[0]);
    }
}
