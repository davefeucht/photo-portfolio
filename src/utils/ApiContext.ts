import { createContext } from 'react';

import { API } from './types';

export const ApiContext = createContext<API | null>(null);
