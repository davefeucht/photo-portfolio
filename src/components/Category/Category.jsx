/*****************
* Category component displays the posts for a particular category.
*****************/

import React from 'react';
import { observer } from 'mobx-react';
import { 
  useParams,
} from 'react-router-dom';
import Posts from '../Posts/Posts.jsx';
import CategoryHeader from '../CategoryHeader/CategoryHeader.jsx';
import './Category.css';

const Category = observer(({ stateStore, api }) => {
  let { categoryId } = useParams();

  api.getPosts(categoryId);
  api.getCategoryInfo(categoryId);

  return ( 
    <div className="category">
      <CategoryHeader stateStore={stateStore} categoryData={stateStore.currentCategoryData}/>
      <Posts stateStore={stateStore} categoryId={categoryId} api={api} />
    </div>
  );
});

Category.displayName = 'Category';

export default Category;