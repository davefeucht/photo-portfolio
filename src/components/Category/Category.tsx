/** ***************
* Category component displays the posts for a particular category.
**************** */

import './Category.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect } from 'react';
import {
    Outlet,
    useParams
} from 'react-router-dom';

import { getCategoryInfo, getPosts, getPostThumbnail } from '../../utils/Api';
import {
    Category as CategoryState,
    ImageData,
    Post as PostState,
    ScreenInfo,
    SiteInfo
} from '../../utils/types';
import Posts from '../Posts/Posts';
import SectionHeader from '../SectionHeader/SectionHeader';

interface CategoryProps {
    maxItemsPerPage: number,
    siteInfo: SiteInfo,
    screenInfo: ScreenInfo,
    currentCategoryPosts: PostState[],
    currentCategoryData: CategoryState,
    setCategoryPosts: (a: PostState[]) => void,
    setCategoryData: (a: CategoryState) => void,
    setThumbnailImageUrl: (a: ImageData) => void
}

const Category: React.FC<CategoryProps> = ({
    maxItemsPerPage,
    siteInfo,
    screenInfo,
    currentCategoryPosts,
    currentCategoryData,
    setCategoryPosts,
    setCategoryData,
    setThumbnailImageUrl
}) => {
    const { categoryId = '' } = useParams();

    useEffect(() => {
        if (categoryId) {
            getPosts(parseInt(categoryId), siteInfo.siteUrl)
                .then((posts: PostState[]) => {
                    setCategoryPosts(posts);
                    currentCategoryPosts.forEach((post, index) => {
                        getPostThumbnail(post.featured_media, siteInfo.siteUrl)
                            .then((thumbUrl: string) => {
                                setThumbnailImageUrl({ post_index: index, image_url: thumbUrl });
                            });
                    });
                });
            getCategoryInfo(parseInt(categoryId), siteInfo.siteUrl)
                .then((categoryInfo: CategoryState) => {
                    setCategoryData(categoryInfo);
                });
        }
    }, [categoryId]);

    return (
        <div className="category">
            <SectionHeader title={currentCategoryData.name} />
            <Posts
                maxItemsPerPage={maxItemsPerPage}
                screenInfo={screenInfo}
                categoryId={parseInt(categoryId)}
                currentCategoryPosts={currentCategoryPosts}
            />
            <Outlet />
        </div>
    );
};

Category.displayName = 'Category';

export default observer(Category);
