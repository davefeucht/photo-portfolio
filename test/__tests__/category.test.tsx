import { render, waitFor } from '@testing-library/react';
import Category from 'components/Category/Category';
import * as React from 'react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import StateStore from 'StateStore/store';
import { ApiContext } from 'utils/ApiContext';
import WordpressAPI from 'utils/WordpressAPI';

const screenInfo = {
    width: 500,
    height: 500
};

jest.mock('../../utils/WordpressAPI');

const setCategoryPostsMock = jest.fn();
const setCategoryDataMock = jest.fn();
const setThumbnailImageUrlMock = jest.fn();

test('Category displays', async () => {
    const store = new StateStore();
    const api = new WordpressAPI(store.siteInfo.siteUrl);
    const posts = await api.getPosts(35);
    const category = await api.getCategoryInfo(35);
    const { container } = render(
        <MemoryRouter initialEntries={['/category/35']}>
            <ApiContext.Provider value={api}>
                <Routes>
                    <Route
                        path="category/:categoryId"
                        element={(
                            <Category
                                maxItemsPerPage={2}
                                screenInfo={screenInfo}
                                currentCategoryPosts={posts}
                                currentCategoryData={category}
                                setCategoryPosts={setCategoryPostsMock}
                                setCategoryData={setCategoryDataMock}
                                setThumbnailImageUrl={setThumbnailImageUrlMock}
                            />
                        )}
                    />
                </Routes>
            </ApiContext.Provider>
        </MemoryRouter>
    );
    await waitFor(() => {
        expect(setCategoryPostsMock).toHaveBeenCalledTimes(1);
        expect(setCategoryDataMock).toHaveBeenCalledTimes(1);
        expect(setThumbnailImageUrlMock).toHaveBeenCalledTimes(2);
    });
    expect(container.firstChild).toMatchSnapshot();
});
