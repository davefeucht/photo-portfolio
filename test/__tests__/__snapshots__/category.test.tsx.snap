// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category displays 1`] = `
<div
  class="category"
>
  <div
    class="section-header"
  >
    <div
      class="title"
    >
      Some other category
    </div>
  </div>
  <div
    class="posts"
  >
    <a
      href="/category/35/post/887"
    >
      <div
        class="post-thumbnail"
        style="background-image: url(https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg);"
      />
    </a>
    <a
      href="/category/35/post/889"
    >
      <div
        class="post-thumbnail"
        style="background-image: url(https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg);"
      />
    </a>
    <div
      class="pagination-navigation"
    >
      <div
        class="pagination-navigation__content"
      >
        <div
          aria-label="pagination"
          class="pagination-navigation__item selected"
        >
          1
        </div>
      </div>
    </div>
  </div>
</div>
`;
